include/openPMD/Datatype.hpp

```diff
@@ -277,161 +279,30 @@ inline constexpr Datatype determineDatatype()
-    {
-        return DT::FLOAT;
-    }
-    else if (decay_equiv<T, double>::value)
-    {
-        return DT::DOUBLE;
-    }
-    else if (decay_equiv<T, long double>::value)
-    {
-        return DT::LONG_DOUBLE;
-    }
-    else if (decay_equiv<T, std::complex<float>>::value)
-    {
-        return DT::CFLOAT;
-    }
-    else if (decay_equiv<T, std::complex<double>>::value)
-    {
-        return DT::CDOUBLE;
-    }
-    else if (decay_equiv<T, std::complex<long double>>::value)
-    {
-        return DT::CLONG_DOUBLE;
-    }
-    else if (decay_equiv<T, std::string>::value)
-    {
-        return DT::STRING;
-    }
-    else if (decay_equiv<T, std::vector<char>>::value)
-    {
-        return DT::VEC_CHAR;
-    }
-    else if (decay_equiv<T, std::vector<short>>::value)
-    {
-        return DT::VEC_SHORT;
-    }
-    else if (decay_equiv<T, std::vector<int>>::value)
-    {
-        return DT::VEC_INT;
-    }
-    else if (decay_equiv<T, std::vector<long>>::value)
-    {
-        return DT::VEC_LONG;
-    }
-    else if (decay_equiv<T, std::vector<long long>>::value)
-    {
-        return DT::VEC_LONGLONG;
-    }
-    else if (decay_equiv<T, std::vector<unsigned char>>::value)
-    {
-        return DT::VEC_UCHAR;
-    }
-    else if (decay_equiv<T, std::vector<signed char>>::value)
-    {
-        return DT::VEC_SCHAR;
-    }
-    else if (decay_equiv<T, std::vector<unsigned short>>::value)
-    {
-        return DT::VEC_USHORT;
-    }
-    else if (decay_equiv<T, std::vector<unsigned int>>::value)
-    {
-        return DT::VEC_UINT;
-    }
-    else if (decay_equiv<T, std::vector<unsigned long>>::value)
-    {
-        return DT::VEC_ULONG;
-    }
-    else if (decay_equiv<T, std::vector<unsigned long long>>::value)
-    {
-        return DT::VEC_ULONGLONG;
-    }
-    else if (decay_equiv<T, std::vector<float>>::value)
-    {
-        return DT::VEC_FLOAT;
-    }
-    else if (decay_equiv<T, std::vector<double>>::value)
-    {
-        return DT::VEC_DOUBLE;
-    }
-    else if (decay_equiv<T, std::vector<long double>>::value)
-    {
-        return DT::VEC_LONG_DOUBLE;
-    }
-    else if (decay_equiv<T, std::vector<std::complex<float>>>::value)
-    {
-        return DT::VEC_CFLOAT;
-    }
-    else if (decay_equiv<T, std::vector<std::complex<double>>>::value)
-    {
-        return DT::VEC_CDOUBLE;
-    }
-    else if (decay_equiv<T, std::vector<std::complex<long double>>>::value)
-    {
-        return DT::VEC_CLONG_DOUBLE;
-    }
-    else if (decay_equiv<T, std::vector<std::string>>::value)
-    {
-        return DT::VEC_STRING;
-    }
-    else if (decay_equiv<T, std::array<double, 7>>::value)
+    return determineDatatype<T>();
+}
+
+template <typename T>
+inline constexpr Datatype determineDatatypeRaw(T const *)
+{
+    return determineDatatype<T>();
+}
+
+template <typename T_ContiguousContainer>
+inline constexpr Datatype determineDatatypeContiguous(T_ContiguousContainer &&)
+{
+    using T_ContiguousContainer_stripped =
```

Comment:
For `determineDatatype`, I tend to just use one API function name with overloads for pointers and cont. containers.
There is no lifetime question here and additional naming is not really needed, adding unnecessary API calls to learn.

What do you think?

Reply:
The different instances would clash:

```cpp
determineDatatype<std::vector<float>>()
```

Should this resolve to:

1. `Datatype::VECTOR_FLOAT` (for a vector-type attribute, e.g. `gridSpacing`)
2. `Datatype::FLOAT` (we have a `std::vector<float>` that we want to store as a chunk)

Both are legitimate. 

Until now, we used:
```cpp
determineDatatype(shareRaw(data))
```

After deprecating `shareRaw`, how should this be rewritten? Without any new APIs, it would be:
```cpp
determineDatatype(*data); // if it is a raw pointer
determineDatatype(data.data()); // if it is a container

// WRONG:
determineDatatype(data); // would resolve to the vector type
```

Further overloading `determineDatatype` does not work, due to the issues above, so it's either the workaround above, or explicit API calls for these instances. (And the whole point of this PR is to replace `shareRaw` with explicit API calls)

Reply:
Checking this again, `determineDatatypeContiguous` is the only place where it is strictly *necessary* to have a new API instead of just an overload.
For `store/loadChunkRaw()`, it's just a good design decision imo (to indicate non-ownership of data), and for `determineDatatypeRaw` it's more a matter of consistency then.

Reply:
I checked the PR again, and we could theoretically do the entire thing only with new overloads, instead of new API calls (even replace `storeChunkRaw` with `storeChunk`). So this is entirely a desing decision:

* `storeChunkRaw()`: I like making it explicit that we are dealing with non-owned, raw memory
* `determineDatatypeContiguous()`: Avoids confusion, as described above
* `determineDatatypeRaw()`: Would then be consistent with the change above

Reply:
I have a slide tendency to overloads, simply because it's C++ pattern (vs. C).

Reply:
For this case, we could make it consistently evaluate to `Datatype::VECTOR_FLOAT`:
```C++
determineDatatype(std::vector<float>());
```
and have the user pass something like `::value_type` to get the fundamental type.